src/raft.rs

```diff
@@ -1079,6 +1048,55 @@ impl<T: Storage> Raft<T> {
         Ok(())
     }
 
+    fn hup(&mut self, transfer_leader: bool) {
+        if self.state == StateRole::Leader {
+            debug!("{} ignoring MsgHup because already leader", self.tag);
+            return;
+        }
+
+        // If there is a pending snapshot, its index will be returned by
+        // `maybe_first_index`. Note that snapshot updates configuration
+        // already, so as long as pending entries don't contain conf change
+        // it's safe to start campaign.
+        let first_index = match self.raft_log.unstable.maybe_first_index() {
+            Some(idx) => idx,
+            None => self.raft_log.applied + 1,
+        };
+
+        let ents = self
+            .raft_log
+            .slice(first_index, self.raft_log.committed + 1, raft_log::NO_LIMIT)
+            .unwrap_or_else(|e| {
+                panic!(
+                    "{} unexpected error getting unapplied entries [{}, {}): {:?}",
+                    self.tag,
+                    first_index,
+                    self.raft_log.committed + 1,
+                    e
+                );
+            });
+        let n = self.num_pending_conf(&ents);
+        if n != 0 && self.raft_log.committed > self.raft_log.applied {
```

Comment:
I think `n != 0` also means `committed > applied`.

Reply:
In this case I think `raft_log.applied` must be less than `snapshot.index`, so I think only `n != 0` is ok here.

Reply:
`raft_log.applied` can be any value less than the snapshot index.